vip_hci/hci_dataset.py

```diff
@@ -1215,3 +1314,149 @@ def subsample(self, window, mode='mean'):
+
+            # save the dataset
+            dataset.save("dataset-1")
+
+            # the following files are created:
+            #   - `./dataset-1-np.vipdata.npz`
+            #   - `./dataset-1-pcl.vipdata`
+
+            # you can restore the object using
+
+            d2 = HCIDataset.from_file("dataset-1")
+
+            # `d2` contains now the same information as `dataset`, e.g. the
+            # actions you performed, the data, ...
+
+        Notes
+        -----
+        If you want to save just the non-array data (no ``cube``, ...), use the
+        ``pickle`` methods directly, e.g.:
+
+        .. code-block:: python
+
+            import pickle
+            with open("my-data-pcl.vipdata", "wb") as f:
+                pickle.dump(dataset, f)
+
+        The objects listed in ``HCIDataset.SAVE_WITH_NP`` are skipped when
+        saving with ``pickle``.
+
+        """
+        
+        # save large arrays with numpy
+        np.savez(_mk_fn(path_prefix, "np"),
+                 **{x:getattr(self, x) for x in self.SAVE_WITH_NP})
+
+        # save rest of object with pickle. calls __getstate__()
+        with open(_mk_fn(path_prefix, "pcl"), "wb") as f:
+            pickle.dump(self, f)
+
+
+    def copy(self):
+        """
+        Creates a new instance of the dataset.
+
+        behaves like `copy.copy()`, but does not ignore self.SAVE_WITH_NP
+        """
+        
+        new_obj = copy.copy(self)
+        for k in self.SAVE_WITH_NP:
+            if k in self.__dict__:
+                new_obj.__dict__[k] = self.__dict__[k]
+
+        return new_obj
+
+    def __getstate__(self):
+        """
+        Used by ``pickle``. Filters out the attributes listed in
+        ``self.SAVE_WITH_NP``.
+
+        Warning: `copy.copy()` also uses this function, so the ``SAVE_WITH_NP``
+        attributes get missing. Use `self.copy()` instead!
+        """
+        # TODO: use `from six import iteritems; for k,v in iteritems(...)` 
+        state = {k:self.__dict__[k] for k in self.__dict__
+                                                  if k not in self.SAVE_WITH_NP}
+        return state
+
+    @classmethod
+    def from_file(cls, path):
+        """
+        Loads an HCIDataset from a given file. See ``save()`` for an example.
+
+        Parameters
+        ----------
+        path : str
+            The path to load the HCIDataset from. Should be of the same form as
+            the ``path`` which was used when ``save()``ing. The suffix
+            ``.vipdata`` is automatically appended.
+
+        Returns
+        -------
+        obj : HCIDataset object
+
+        """
+        with open(_mk_fn(path, "pcl"), "rb") as f:
+            obj = pickle.load(f)
+
+        np_stuff_dict = np.load(str(_mk_fn(path, "np"))+".npz")
+        # np always adds `.npz` to the filename...
+
+        for k in np_stuff_dict:
+            data = np_stuff_dict[k]
+            setattr(obj, k, data)
+
+        return obj
+
+
+
+
```

Comment:
What is the purpose of all these empty lines?

Reply:
That is a visual hint for myself, that there is a function at the bottom of the file which may be moved to another module 😋